src/signature.c

```diff
@@ -2009,3 +2049,274 @@ const OSSL_DISPATCH p11prov_ecdsa_signature_functions[] = {
+        return RET_OSSL_ERR;
+    }
+
+    ret =
+        p11prov_sig_operate(sigctx, sig, siglen, sigsize, (void *)tbs, tbslen);
+    if (ret != CKR_OK) {
+        return RET_OSSL_ERR;
+    }
+    return RET_OSSL_OK;
+}
+
+static int p11prov_eddsa_digest_verify_init(void *ctx, const char *digest,
+                                            void *provkey,
+                                            const OSSL_PARAM params[])
+{
+    CK_RV ret;
+
+    P11PROV_debug("eddsa digest verify init (ctx=%p, key=%p, params=%p)", ctx,
+                  provkey, params);
+
+    if (digest != NULL && digest[0] != '\0') {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
+        return RET_OSSL_ERR;
+    }
+
+    ret = p11prov_sig_op_init(ctx, provkey, CKF_VERIFY, digest);
+    if (ret != CKR_OK) {
+        return RET_OSSL_ERR;
+    }
+
+    return p11prov_eddsa_set_ctx_params(ctx, params);
+}
+
+static int p11prov_eddsa_digest_verify(void *ctx, const unsigned char *sig,
+                                       size_t siglen, const unsigned char *tbs,
+                                       size_t tbslen)
+{
+    P11PROV_SIG_CTX *sigctx = (P11PROV_SIG_CTX *)ctx;
+    CK_RV ret;
+
+    P11PROV_debug("eddsa digest verify (ctx=%p, tbs=%p, tbslen=%zu)", ctx, tbs,
+                  tbslen);
+
+    if (sigctx == NULL) {
+        return RET_OSSL_ERR;
+    }
+
+    ret = p11prov_sig_operate(sigctx, (void *)sig, NULL, siglen, (void *)tbs,
+                              tbslen);
+    if (ret != CKR_OK) {
+        return RET_OSSL_ERR;
+    }
+    return RET_OSSL_OK;
+}
+
+static int p11prov_eddsa_get_ctx_params(void *ctx, OSSL_PARAM *params)
+{
+    P11PROV_SIG_CTX *sigctx = (P11PROV_SIG_CTX *)ctx;
+    OSSL_PARAM *p;
+    int ret;
+
+    /* todo sig params:
+        OSSL_SIGNATURE_PARAM_ALGORITHM_ID
+     */
+
+    P11PROV_debug("eddsa get ctx params (ctx=%p, params=%p)", ctx, params);
+
+    p = OSSL_PARAM_locate(params, OSSL_SIGNATURE_PARAM_ALGORITHM_ID);
+    if (p) {
+        if (sigctx->mechtype != CKM_EDDSA) {
+            return RET_OSSL_ERR;
+        }
+        CK_ULONG size = p11prov_obj_get_key_bit_size(sigctx->key);
+        switch (size) {
+        case ED25519_BIT_SIZE:
+            ret = OSSL_PARAM_set_octet_string(p, der_ed25519_algorithm_id,
+                                              sizeof(der_ed25519_algorithm_id));
+            break;
+        case ED448_BIT_SIZE:
+            ret = OSSL_PARAM_set_octet_string(p, der_ed448_algorithm_id,
+                                              sizeof(der_ed448_algorithm_id));
+            break;
+        default:
+            ret = RET_OSSL_ERR;
+            break;
+        }
+        if (ret != RET_OSSL_OK) {
+            return ret;
+        }
+    }
+
+    return RET_OSSL_OK;
+}
+
+#ifndef OSSL_SIGNATURE_PARAM_INSTANCE
+#define OSSL_SIGNATURE_PARAM_INSTANCE "instance"
+#endif
+#ifndef OSSL_SIGNATURE_PARAM_CONTEXT_STRING
+#define OSSL_SIGNATURE_PARAM_CONTEXT_STRING "context-string"
+#endif
```

Comment:
Are these missing in some older version of openssl than 3.0.7? If so, if we bump the required version we can drop these. If there is different reason to include them, comment saying why and how long they are needed would be worth placing here.

Reply:
They are missing in 3.0.7 (they are upstream only)

Reply:
We've done this for other stuff, I am fine with keeping these until they are not necessary anymore, they are harmless.